text/0000-nonparametric-dropck.md

```diff
@@ -0,0 +1,541 @@
+
+The Drop Check rule (`dropck`) for [Sound Generic Drop][] relies on a
+reasoning process that needs to infer that the behavior of a
+polymorphic function (e.g. `fn foo<T>`) does not depend on the
+concrete type instantiations of any of its *unbounded* type parameters
+(e.g. `T` in `fn foo<T>`), at least beyond the behavior of the
+destructor (if any) for those type parameters.
+
+[Sound Generic Drop]: https://github.com/rust-lang/rfcs/blob/master/text/0769-sound-generic-drop.md
+
+This property is a (weakened) form of a property known in academic
+circles as *Parametricity*.
+(See e.g. [Reynolds, IFIP 1983][Rey83], [Wadler, FPCA 1989][Wad89].)
+
+ * Parametricity, in this context, essentially says that the compiler
+   can reason about the body of `foo` (and the subroutines that `foo`
+   invokes) without having to think about the particular concrete
+   types that the type parameter `T` is instantiated with.
+   `foo` cannot do anything with a `t: T` except:
+
+   1. move `t` to some other owner expecting a `T` or,
+
+   2. drop `t`, running its destructor and freeing associated resources.
+
+ * For example, this allows the compiler to deduce that even if `T` is
+   instantiated with a concrete type like `&Vec<u32>`, the body of
+   `foo` cannot actually read any `u32` data out of the vector. More
+   details about this are available on the [Sound Generic Drop][] RFC.
+
+## "Mistakes were made"
+
+The parametricity-based reasoning in the
+[Drop Check analysis][Sound Generic Drop] (`dropck`) was clever, but
+fragile and unproven.
+
+ * Regarding its fragility, it has been shown to have
+   [bugs][parametricity-insufficient]; in particular, parametricity is
+   a necessary but *not* sufficient condition to justify the
+   inferences that `dropck` makes.
+
+ * Regarding its unproven nature, `dropck` violated the heuristic in
+   Rust's design to not incorporate ideas unless those ideas had
+   already been proven effective elsewhere.
+
+[parametricity-insufficient]: https://github.com/rust-lang/rust/issues/26656
+
+These issues might alone provide motivation for ratcheting back on
+`dropck`'s rules in the short term, putting in a more conservative
+rule in the stable release channel while allowing experimentation with
+more-aggressive feature-gated rules in the development nightly release
+channel.
+   
+However, there is also a specific reason why we want to ratchet back
+on the `dropck` analysis as soon as possible.
+
+## Impl specialization is inherently non-parametric
+
+The parametricity requirement in the Drop Check rule over-restricts
+the design space for future language changes.
+
+In particular, the [impl specialization] RFC describes a language
+change that will allow the invocation of a polymorphic function `f` to
+end up in different sequences of code based solely on the concrete
+type of `T`, *even* when `T` has no trait bounds within its
+declaration in `f`.
+
+[impl specialization]: https://github.com/rust-lang/rfcs/pull/1210
+
```

Comment:
On a slightly more concrete note: One of the consequences of parametricity is that if you have a function `fn foo<T>(t: T) -> T`, then there is literally only one thing the function can do: Return its argument. (It would also mutate some other state, but that's irrelevant here.) In particular, the function can not dereference any pointer contained in `t`, since it can't even know whether there are pointers in `t`! This is because `foo` has to do "basically the same thing" for all `T`. After all, there's only one `foo` implementing all that!

However, once we have specialization, one could write a special version of this function for `T = u32` that increments its argument. Or a special version for `T = &mut i32` that increments the data in the pointer, and then return the pointer. Suddenly, the type of the function alone no longer guarantees that the function has to do "basically the same thing" for all `T`.